### PR TITLE
[sailfishos][embedlite] Allow EmbedLiteApp window re-creation. Contributes to JB#54813

### PR DESCRIPTION
**Pull request by paveltkv**
The full set of PRs related to this:
1. https://github.com/sailfishos/gecko-dev/pull/9 (this)
2. https://github.com/sailfishos/qtmozembed/pull/3
3. https://github.com/sailfishos/sailfish-browser/pull/897

### CODE REVIEW

File: embedding/embedlite/embedshared/EmbedLiteWindowBaseChild.h [hunk 1]
```diff
@@ -40,11 +42,15 @@ class EmbedLiteWindowBaseChild : public PEmbedLiteWindowChild
   void CreateWidget();
 
   uint32_t mId;
+  EmbedLiteWindowListener *const mListener;
   nsCOMPtr<nsIWidget> mWidget;
   LayoutDeviceIntRect mBounds;
   mozilla::ScreenRotation mRotation;
   RefPtr<CancelableRunnable> mCreateWidgetTask;
 
+  bool initialized;
```
Comment (rainemak) [inline embedding/embedlite/embedshared/EmbedLiteWindowBaseChild.h:51]:
```suggestion
  bool mInitialized;
```

File: embedding/embedlite/embedshared/EmbedLiteWindowBaseChild.h [hunk 2]
```diff
@@ -40,11 +42,15 @@ class EmbedLiteWindowBaseChild : public PEmbedLiteWindowChild
...
+  bool destroyAfterInit;
```
Comment (rainemak) [inline embedding/embedlite/embedshared/EmbedLiteWindowBaseChild.h:52]:
```suggestion
  bool mDestroyAfterInit;
```

File: embedding/embedlite/embedshared/EmbedLiteWindowBaseChild.cpp [hunk 3]
```diff
@@ -182,7 +190,14 @@ void EmbedLiteWindowBaseChild::CreateWidget()
...
               );
   GetWidget()->UpdateSize();
+
+  initialized = true;
+
   Unused << SendInitialized();
+
+  if (destroyAfterInit) {
+    RecvDestroy();
+  }
```
Comment (rainemak) [inline embedding/embedlite/embedshared/EmbedLiteWindowBaseChild.cpp:200]: This doesn't look like right order... Very likely destroying here and sending initialized can lead to bad state.

Without checking the whole init view code path down to the browser side, following should be safer:
```
if (mDestroyAfterInit) {
    RecvDestroy();
} else {
    Unused << SendInitialized();
}
```
Does it even make sense to create nsWindow if mDestroyAfterInit is true, looks suspicious? 

Right after mCreateWidgetTask clearing... I think you could do:

```
if (mDestroyAfterInit) {
    RecvDestroy();
    return;
}
```

File: embedding/embedlite/embedshared/EmbedLiteWindowBaseChild.cpp [hunk 4]
```diff
@@ -89,7 +91,13 @@ void EmbedLiteWindowBaseChild::ActorDestroy(ActorDestroyReason aWhy)
 
 mozilla::ipc::IPCResult EmbedLiteWindowBaseChild::RecvDestroy()
 {
+  if (initialized == false) {
+    destroyAfterInit = true;
+    return IPC_OK();
+  }
+
   LOGT("destroy");
+  mWidget->Destroy();
```
Comment (rainemak) [inline embedding/embedlite/embedshared/EmbedLiteWindowBaseChild.cpp:100]: Given that https://github.com/sailfishos/gecko-dev/pull/9/files#r706130282 we'd return earlier on the CreateWidget then this should be guarded with

```
if (mWidget) {
    mWidget->Destroy();
}
```

File: embedding/embedlite/embedshared/EmbedLiteWindowBaseChild.cpp [hunk 5]
```diff
@@ -89,7 +91,13 @@ void EmbedLiteWindowBaseChild::ActorDestroy(ActorDestroyReason aWhy)
 
 mozilla::ipc::IPCResult EmbedLiteWindowBaseChild::RecvDestroy()
 {
+  if (initialized == false) {
```
Comment (rainemak) [inline embedding/embedlite/embedshared/EmbedLiteWindowBaseChild.cpp:94]:
```suggestion
  if (!mInitialized) {
```

File: embedding/embedlite/embedshared/EmbedLiteViewBaseChild.cpp [hunk 6]
```diff
@@ -115,6 +115,8 @@ EmbedLiteViewBaseChild::EmbedLiteViewBaseChild(const uint32_t& aWindowId, const
   , mVirtualKeyboardHeight(0)
   , mIMEComposing(false)
   , mPendingTouchPreventedBlockId(0)
+  , initialized(false)
+  , destroyAfterInit(false)
```
Comment (rainemak) [inline embedding/embedlite/embedshared/EmbedLiteViewBaseChild.cpp:119]:
```suggestion
  , mDestroyAfterInit(false)
```

Same for mInitialized

File: embedding/embedlite/EmbedLiteApp.cpp [hunk 7]
```diff
@@ -471,15 +476,19 @@ EmbedLiteApp::CreateView(EmbedLiteWindow* aWindow, uint32_t aParent, bool aIsPri
...
 EmbedLiteWindow*
-EmbedLiteApp::CreateWindow(int width, int height)
+EmbedLiteApp::CreateWindow(int width, int height, EmbedLiteWindowListener *aListener)
 {
   LOGT();
   NS_ASSERTION(mState == INITIALIZED, "The app must be up and runnning by now");
   static uint32_t sWindowCreateID = 0;
   sWindowCreateID++;
 
+  if (aListener == nullptr) {
```
Comment (rainemak) [inline embedding/embedlite/EmbedLiteApp.cpp:486]:
```suggestion
  if (!aListener) {
```

**Review by rainemak — APPROVED** [review]

**Review by okodron — APPROVED** [review]

**Pull request merged**